### PR TITLE
refactor: replace `[a; b]` syntax with `$ARRAY_REPEAT$`

### PR DESCRIPTION
**Pull request by jaschdoc**
- [x] Remove option type from `ArrayNew` and the corresponding `optional(... ~ Expression ~ ...)`

Related to https://github.com/flix/flix/issues/4760

### CODE REVIEW

**Comment by jaschdoc** [comment]
Just community build that fails due to http

File: main/test/flix/Test.Exp.Regions.flix [hunk 1]
```diff
@@ -80,7 +84,7 @@ namespace Test/Exp/Regions {
     @test
     def testRegions1Region01(): Bool =
         region r {
-            let a = [1; 8];
+            let a = $ARRAY_REPEAT$(r, 1, 8);
```
Comment (magnus-madsen) [inline main/test/flix/Test.Exp.Regions.flix:87]: Can we use `Array.new` in this file?

File: main/test/flix/Test.Exp.Regions.flix [hunk 2]
```diff
@@ -298,10 +302,10 @@ namespace Test/Exp/Regions {
     @test
     def testRegions2Regions04(): Bool \ IO =
         region r1 {
-            let a1 = [1; 8] @ r1;
+            let a1 = $ARRAY_REPEAT$(r1, 1, 8);
             a1[0] == 1 and 1 ==
             region _ {
-                let a2 = [1; 8] @ Static;
+                let a2 = $ARRAY_REPEAT$(unsafe_cast magic() as Region[false], 1, 8);
```
Comment (magnus-madsen) [inline main/test/flix/Test.Exp.Regions.flix:308]: This file is supposed to test regions, not all kinds of ARRAY_REPEAT, unsafe cast, and magic stuff! :)

File: main/src/library/Array.flix [hunk 3]
```diff
@@ -66,12 +66,10 @@ namespace Array {
...
 
     ///
-    /// Returns a fresh array of length `l` filled with the element `x`.
-    ///
-    /// Equivalent to the expression `[x; l]`.
+    /// Returns a fresh array of length `l` filled with the element `x` in the region r.
     ///
     pub def new(r: Region[r], x: a, l: Int32): Array[a, r] \ Write(r) =
-        [x; l] @ r
+        $ARRAY_REPEAT$(r, x, l)
```
Comment (magnus-madsen) [inline main/src/library/Array.flix:72]: Ok, so I like the name new. But then lets call it `$ARRAY_NEW$`.

File: main/test/flix/Test.Exp.Regions.flix (hunk 2 above)
Comment (jaschdoc) [reply main/test/flix/Test.Exp.Regions.flix:308]: Yeah but our safety rules are sort of ruining things here / making things complicated.

$ARRAY_REPEAT$ requires a term level value that carries region information. So naturally you would unsafe_cast unit to some region type but that is rejected as a safety error. Then we have to introduce a function whose return type is a type variable so the compiler accepts an unsafe_cast from a flix primitive type to a type variable.

This is all just to allocate memory in the static region.

Comment (jaschdoc) [reply main/test/flix/Test.Exp.Regions.flix:308]: All just to do the equivalent of

[1;8] @ Static

Also, $ARRAY_REPEAT$(false, ...) doesn't work as false is a Boolean not a region (just stating the obvious here)

Comment (magnus-madsen) [reply main/test/flix/Test.Exp.Regions.flix:308]: Can we just use a normal region?

File: main/test/flix/Test.Exp.Regions.flix (hunk 1 above)
Comment (jaschdoc) [reply main/test/flix/Test.Exp.Regions.flix:87]: I suppose we could. Just wanted to reduce dependency on the std lib but if we're already using mutmap and mutlist then we may as well just use array

File: main/test/flix/Test.Exp.Regions.flix [hunk 4]
```diff
@@ -1,5 +1,9 @@
 namespace Test/Exp/Regions {
 
+    def magic(): a = typematch () {
```
[the remark this answers is not in the document]
Comment (jaschdoc) [reply main/test/flix/Test.Exp.Regions.flix:3]: See comment below :)

File: main/test/flix/Test.Exp.Regions.flix (hunk 2 above)
Comment (magnus-madsen) [reply main/test/flix/Test.Exp.Regions.flix:308]: We are trying to get rid of static anyway

File: main/test/flix/Test.Exp.Regions.flix (hunk 4 above)
[the remark this answers is not in the document]
Comment (jaschdoc) [reply main/test/flix/Test.Exp.Regions.flix:3]: I agree that this is very bad 😃

File: main/test/ca/uwaterloo/flix/library/TestRandom.flix [hunk 5]
```diff
@@ -2,15 +2,15 @@
 def testChoose01(): Option[Int32] \ NonDet =
     region r {
         let rnd = Random.new();
-        let arr = [42i32; 10] @ r;
+        let arr = $ARRAY_REPEAT$(r, 42i32, 10);
```
[the remark this answers is not in the document]
Comment (jaschdoc) [reply main/test/ca/uwaterloo/flix/library/TestRandom.flix:5]: Yes my bad

File: main/test/flix/Test.Exp.Regions.flix (hunk 1 above)
Comment (magnus-madsen) [reply main/test/flix/Test.Exp.Regions.flix:87]: Great

File: main/test/flix/Test.Exp.Regions.flix (hunk 2 above)
Comment (jaschdoc) [reply main/test/flix/Test.Exp.Regions.flix:308]: Ah perfect! That's much easier haha

**Comment by jaschdoc** [comment]
I've kept the `Static` region for now so the tests are equivalent to what we had before.

**Comment by magnus-madsen** [comment]
Thanks!

**Pull request merged**